File: msr/reference/zig/balanced-parens.zig
pub fn is_balanced(s: []const u8) bool {
    var stack: [256]u8 = undefined;
    var top: usize = 0;
    for (s) |c| {
        switch (c) {
            '(', '[', '{' => {
                if (top == stack.len) return false;
                stack[top] = switch (c) {
                    '(' => ')',
                    '[' => ']',
                    else => '}',
                };
                top += 1;
            },
            ')', ']', '}' => {
                if (top == 0 or stack[top - 1] != c) return false;
                top -= 1;
            },
            else => {},
        }
    }
    return top == 0;
}

File: msr/reference/go/balanced-parens.go
package main

func is_balanced(s string) bool {
	closer := map[rune]rune{'(': ')', '[': ']', '{': '}'}
	stack := []rune{}
	for _, c := range s {
		if cl, ok := closer[c]; ok {
			stack = append(stack, cl)
		} else if c == ')' || c == ']' || c == '}' {
			if len(stack) == 0 || stack[len(stack)-1] != c {
				return false
			}
			stack = stack[:len(stack)-1]
		}
	}
	return len(stack) == 0
}
